app/controllers/CancelLeaveSchemeController.scala

```diff
@@ -41,39 +43,53 @@ class CancelLeaveSchemeController @Inject()(
                                              getData: DataRetrievalAction,
                                              formProvider: CancelLeaveSchemeFormProvider,
                                              config: FrontendAppConfig,
+                                             clock: Clock,
                                              val controllerComponents: MessagesControllerComponents,
                                              view: CancelLeaveSchemeView,
                                              registrationService: RegistrationService
                                            )(implicit ec: ExecutionContext) extends FrontendBaseController with I18nSupport with Logging {
 
   val form = formProvider()
 
-  def onPageLoad(waypoints: Waypoints): Action[AnyContent] = (identify andThen getData) {
+  def onPageLoad(waypoints: Waypoints): Action[AnyContent] = (identify andThen getData).async {
     implicit request =>
-      val preparedForm = request.userAnswers.getOrElse(UserAnswers(request.userId)).get(CancelLeaveSchemePage) match {
-        case None => form
-        case Some(value) => form.fill(value)
+      onValidExclusion {
+        val preparedForm = request.userAnswers.getOrElse(UserAnswers(request.userId)).get(CancelLeaveSchemePage) match {
+          case None => form
+          case Some(value) => form.fill(value)
+        }
+        Future.successful(Ok(view(preparedForm, waypoints)))
       }
-
-      Ok(view(preparedForm, waypoints))
   }
 
   def onSubmit(waypoints: Waypoints): Action[AnyContent] = (identify andThen getData).async {
     implicit request =>
+      onValidExclusion {
+        form.bindFromRequest().fold(
+          formWithErrors =>
+            Future.successful(BadRequest(view(formWithErrors, waypoints))),
 
-      form.bindFromRequest().fold(
-        formWithErrors =>
-          Future.successful(BadRequest(view(formWithErrors, waypoints))),
+          hasCancelled => {
+            val originalAnswers: UserAnswers = request.userAnswers.getOrElse(UserAnswers(request.userId))
+            for {
+              updatedAnswers <- Future.fromTry(originalAnswers.set(CancelLeaveSchemePage, hasCancelled))
+              _ <- sessionRepository.set(updatedAnswers)
+              result <- amendRegistration(waypoints, hasCancelled, updatedAnswers)
+            } yield result
+          }
+        )
+      }
+  }
 
-        hasCancelled => {
-          val originalAnswers: UserAnswers = request.userAnswers.getOrElse(UserAnswers(request.userId))
-          for {
-            updatedAnswers <- Future.fromTry(originalAnswers.set(CancelLeaveSchemePage, hasCancelled))
-            _ <- sessionRepository.set(updatedAnswers)
-            result <- amendRegistration(waypoints, hasCancelled, updatedAnswers)
-          } yield result
-        }
-      )
+  private def onValidExclusion(f: => Future[Result])(implicit request: OptionalDataRequest[AnyContent]): Future[Result] = {
```

Comment:
do you not think it would be better to make this into a filter? Or is there a specific reason to make it a method within here?